chimney/src/main/scala/io/scalaland/chimney/internal/compiletime/ResultDefinitions.scala

```diff
@@ -0,0 +1,324 @@
+package io.scalaland.chimney.internal.compiletime
+
+import scala.annotation.{nowarn, tailrec}
+import scala.collection.compat.*
+
+@nowarn("msg=The outer reference in this type test cannot be checked at run time.")
+private[compiletime] trait ResultDefinitions { this: Definitions =>
+
+  sealed protected trait DerivationError extends Product with Serializable
+  protected object DerivationError {
+
+    // TODO: expand as needed
+    final case class MacroException(throwable: Throwable) extends DerivationError
+    final case class NotYetImplemented(functionality: String) extends DerivationError
+  }
+
+  final protected case class DerivationErrors(head: DerivationError, tail: Vector[DerivationError]) {
+
+    def ++(errors: DerivationErrors): DerivationErrors =
+      DerivationErrors(head, tail ++ Vector(errors.head) ++ errors.tail)
+  }
+  protected object DerivationErrors {
+
+    def apply(error: DerivationError, errors: DerivationError*): DerivationErrors =
+      apply(error, errors.toVector)
+
+    def fromException(throwable: Throwable): DerivationErrors =
+      apply(DerivationError.MacroException(throwable), Vector.empty)
+  }
+
+  /** Lazily evaluated log entry */
+  final class LogEntry(val nesting: Int, thunk: => String) {
+    lazy val message: String = thunk
+  }
+
+  /** Contains everything we would like to access in our computations without resorting to globals.
+   *
+   * Possible usages:
+   * - access to config
+   * - appending logs and diagnostics
+   * - storing cache for intermediate results
+   */
+  final case class Context(
+      config: TransformerConfig = TransformerConfig(),
```

Comment:
I'd avoid coupling result monad with just transformer-specific stuff. In the end we may use it for derive patchers too. Wouldn't it be too hard to parameterize the context at this point?